<a href="https://colab.research.google.com/github/JohnMakgakgaDS/ML-intern-Cognifyz/blob/main/mltask2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Task 2: Restaurant Recommendation

Objective: Create a restaurant recommendation
system based on user preferences.

Preprocess the dataset by handling missing
values and encoding categorical variables.

In [1]:
#Load data from the local machine into colab
from google.colab import files

uploaded = files.upload()

Saving Dataset .csv to Dataset .csv


In [2]:
#Libraries or data science packages to be used along the program tasks

import pandas as pd
import numpy as np

In [3]:
# Assuming the uploaded file is in CSV format, load it into a pandas DataFrame
# Replace 'your_file.csv' with the actual file name you uploaded
file_name = list(uploaded.keys())[0]  # Get the uploaded file name
df = pd.read_csv(file_name)

# View the first few rows of the dataset
df.head()

Restaurant ID         Restaurant Name  Country Code              City  \
0        6317637        Le Petit Souffle           162       Makati City   
1        6304287        Izakaya Kikufuji           162       Makati City   
2        6300002  Heat - Edsa Shangri-La           162  Mandaluyong City   
3        6318506                    Ooma           162  Mandaluyong City   
4        6314302             Sambo Kojin           162  Mandaluyong City   

                                             Address  \
0  Third Floor, Century City Mall, Kalayaan Avenu...   
1  Little Tokyo, 2277 Chino Roces Avenue, Legaspi...   
2  Edsa Shangri-La, 1 Garden Way, Ortigas, Mandal...   
3  Third Floor, Mega Fashion Hall, SM Megamall, O...   
4  Third Floor, Mega Atrium, SM Megamall, Ortigas...   

                                     Locality  \
0   Century City Mall, Poblacion, Makati City   
1  Little Tokyo, Legaspi Village, Makati City   
2  Edsa Shangri-La, Ortigas, Mandaluyong City   
3      SM Megamall, Ortigas, Mandaluyong City   
4      SM Megamall, Ortigas, Mandaluyong City   

                                    Locality Verbose   Longitude   Latitude  \
0  Century City Mall, Poblacion, Makati City, Mak...  121.027535  14.565443   
1  Little Tokyo, Legaspi Village, Makati City, Ma...  121.014101  14.553708   
2  Edsa Shangri-La, Ortigas, Mandaluyong City, Ma...  121.056831  14.581404   
3  SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.056475  14.585318   
4  SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.057508  14.584450   

                           Cuisines  ...          Currency Has Table booking  \
0        French, Japanese, Desserts  ...  Botswana Pula(P)               Yes   
1                          Japanese  ...  Botswana Pula(P)               Yes   
2  Seafood, Asian, Filipino, Indian  ...  Botswana Pula(P)               Yes   
3                   Japanese, Sushi  ...  Botswana Pula(P)                No   
4                  Japanese, Korean  ...  Botswana Pula(P)               Yes   

  Has Online delivery Is delivering now Switch to order menu Price range  \
0                  No                No                   No           3   
1                  No                No                   No           3   
2                  No                No                   No           4   
3                  No                No                   No           4   
4                  No                No                   No           4   

   Aggregate rating  Rating color Rating text Votes  
0               4.8    Dark Green   Excellent   314  
1               4.5    Dark Green   Excellent   591  
2               4.4         Green   Very Good   270  
3               4.9    Dark Green   Excellent   365  
4               4.8    Dark Green   Excellent   229  

[5 rows x 21 columns]

#show number of rows and columns

In [4]:
df.shape

(9551, 21)

There are 9551 rows in the dataset
There are 21 columns in the dataset

#check missing values in each column

In [5]:
df.isnull().sum()

Restaurant ID           0
Restaurant Name         0
Country Code            0
City                    0
Address                 0
Locality                0
Locality Verbose        0
Longitude               0
Latitude                0
Cuisines                9
Average Cost for two    0
Currency                0
Has Table booking       0
Has Online delivery     0
Is delivering now       0
Switch to order menu    0
Price range             0
Aggregate rating        0
Rating color            0
Rating text             0
Votes                   0
dtype: int64

There's 9 missing values in column Cuisines

Handling missing values for Cuisines

In [6]:
 #Fill missing values with a placeholder
df['Cuisines'].fillna('Unknown', inplace=True)


<ipython-input-6-ea5a34f7acc5>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Cuisines'].fillna('Unknown', inplace=True)


In [7]:
#checking for validation
df.isnull().sum()

Restaurant ID           0
Restaurant Name         0
Country Code            0
City                    0
Address                 0
Locality                0
Locality Verbose        0
Longitude               0
Latitude                0
Cuisines                0
Average Cost for two    0
Currency                0
Has Table booking       0
Has Online delivery     0
Is delivering now       0
Switch to order menu    0
Price range             0
Aggregate rating        0
Rating color            0
Rating text             0
Votes                   0
dtype: int64

Filled missing values with Unknown , There are no missing values .

#Encoding categorical variables

In [9]:
df['Has Table booking'] = df['Has Table booking'].replace({'Yes': 1, 'No': 0})
df['Has Online delivery'] = df['Has Online delivery'].replace({'Yes': 1, 'No': 0})
df['Is delivering now'] = df['Is delivering now'].replace({'Yes': 1, 'No': 0})
df['Switch to order menu'] = df['Switch to order menu'].replace({'Yes': 1, 'No': 0})
df['Rating text'] = df['Rating text'].replace({'Poor': 0, 'Not rated': 1,'Average':2,'Good':3,'Very Good':4,'Excellent':5 })
df['Rating color'] = df['Rating color'].replace({'Red': 0, 'White': 1,'Orange':2,'Yellow':3,'Green':4,'Dark Green':5 })

<ipython-input-9-b147e7f11672>:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['Has Table booking'] = df['Has Table booking'].replace({'Yes': 1, 'No': 0})
<ipython-input-9-b147e7f11672>:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['Has Online delivery'] = df['Has Online delivery'].replace({'Yes': 1, 'No': 0})
<ipython-input-9-b147e7f11672>:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False

In [17]:
from sklearn.preprocessing import MultiLabelBinarizer

# Assuming 'Cuisines' contains comma-separated lists of cuisines, e.g., "Italian, Chinese"
# First, split the 'Cuisines' column into a list of cuisines
df['Cuisines'] = df['Cuisines'].apply(lambda x: x.split(', ') if isinstance(x, str) else [])

# Initialize the MultiLabelBinarizer
mlb = MultiLabelBinarizer()

# Apply MultiLabelBinarizer to create binary columns for each cuisine
cuisine_encoded = mlb.fit_transform(df['Cuisines'])

# Convert the binary matrix to a DataFrame with the cuisine names as column names
cuisine_encoded_df = pd.DataFrame(cuisine_encoded, columns=mlb.classes_)

# Add the new binary columns back to the original dataframe
df = pd.concat([df, cuisine_encoded_df], axis=1)



In [18]:
#Validation
df.iloc[0:5,10:21]

Average Cost for two          Currency  Has Table booking  \
0             -0.006154  Botswana Pula(P)                  1   
1              0.000049  Botswana Pula(P)                  1   
2              0.173743  Botswana Pula(P)                  1   
3              0.018659  Botswana Pula(P)                  0   
4              0.018659  Botswana Pula(P)                  1   

   Has Online delivery  Is delivering now  Switch to order menu  Price range  \
0                    0                  0                     0     1.319803   
1                    0                  0                     0     1.319803   
2                    0                  0                     0     2.424090   
3                    0                  0                     0     2.424090   
4                    0                  0                     0     2.424090   

   Aggregate rating  Rating color  Rating text  Votes  
0          1.407131             5            5    314  
1          1.209281             5            5    591  
2          1.143331             4            4    270  
3          1.473081             5            5    365  
4          1.407131             5            5    229

Determining the criteria for restaurant recommendations e.g., cuisine preference,price range.



*   Cuisine preference: The user could specify their preferred cuisines (e.g., Italian, Chinese, etc.).
*   Price range: The user could specify a price range (e.g., low, medium, high).
*   Rating: The user could prefer restaurants with higher ratings.
*   Delivery options: Whether the user wants delivery options or not.

In [19]:
# List of features to use in the recommendation system
features = ['Average Cost for two', 'Price range', 'Aggregate rating', 'Has Table booking',
            'Has Online delivery', 'Is delivering now', 'Switch to order menu'] + list(cuisine_encoded_df.columns)

# Create the feature matrix X
X = df[features]


In [23]:
from sklearn.preprocessing import StandardScaler

# Normalize the numerical features (e.g., 'Average Cost for two', 'Price range', etc.)
numerical_features = ['Average Cost for two', 'Price range', 'Aggregate rating']
X[numerical_features] = StandardScaler().fit_transform(X[numerical_features])


<ipython-input-23-3112c3fb6589>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X[numerical_features] = StandardScaler().fit_transform(X[numerical_features])


#Implement Content-Based Filtering

Calculate Similarities

In [24]:
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.preprocessing import StandardScaler

# Normalize numerical features
scaler = StandardScaler()
numerical_features = ['Average Cost for two', 'Price range', 'Aggregate rating']
df[numerical_features] = scaler.fit_transform(df[numerical_features])



# Compute cosine similarity between all restaurants
cosine_sim = cosine_similarity(X)

# Function to recommend restaurants based on a restaurant ID
def recommend_restaurants(restaurant_id, top_n=10):
    # Get the similarity scores for the restaurant
    sim_scores = list(enumerate(cosine_sim[restaurant_id]))

    # Sort the restaurants by similarity score (in descending order)
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    # Get the top n most similar restaurants (excluding the restaurant itself)
    sim_scores = sim_scores[1:top_n+1]

    # Get restaurant indices
    restaurant_indices = [x[0] for x in sim_scores]

    # Return the most similar restaurants
    return df.iloc[restaurant_indices][['Restaurant ID', 'Cuisines', 'Average Cost for two', 'Price range', 'Aggregate rating']]

# Example: Recommend restaurants similar to the restaurant with ID = 5
recommended_restaurants = recommend_restaurants(restaurant_id=5, top_n=10)
print(recommended_restaurants)


      Restaurant ID Cuisines  Average Cost for two  Price range  \
2450          15774       []             -0.006154     1.319803   
2452          15321       []              0.000049     1.319803   
2285        2100870       []              0.006252     1.319803   
6709         312476       []              0.006252     1.319803   
9259        2800903       []              0.006252     1.319803   
844         3000048       []              0.012456     1.319803   
9275        2800019       []              0.012456     1.319803   
745        18339874       []              0.018659     1.319803   
3438           3192       []              0.018659     1.319803   
4386           3190       []              0.018659     1.319803   

      Aggregate rating  
2450          1.143331  
2452          1.143331  
2285          1.143331  
6709          1.143331  
9259          1.143331  
844           1.143331  
9275          1.143331  
745           1.143331  
3438          1.143331  
4386        

Handling Cuisine Preferences

In [28]:
def recommend_by_cuisine(cuisine_preference, top_n=10):
    # Filter the dataset by cuisine preference using multi-hot encoding
    # Check if the cuisine exists in the multi-hot encoded features
    filtered_data = df[df[cuisine_preference] == 1]

    # If no matching restaurants, return an empty dataframe or an alternative recommendation
    if filtered_data.empty:
        return pd.DataFrame()  # Return an empty DataFrame if no restaurants are found

    # Compute cosine similarity for the filtered restaurants
    X_filtered = filtered_data[features]
    cosine_sim_filtered = cosine_similarity(X_filtered)

    # Recommend top N similar restaurants based on the filtered data
    # We calculate cosine similarity for all restaurants, but we only use the filtered ones for recommendations
    sim_scores = list(enumerate(cosine_sim_filtered[0]))  # For the first restaurant in the filtered set
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)[1:top_n+1]  # Get top N similar ones

    # Get the indices of the recommended restaurants
    recommended_indices = [x[0] for x in sim_scores]

    # Return the most similar restaurants from the filtered data
    return filtered_data.iloc[recommended_indices][['Restaurant ID', 'Cuisines', 'Average Cost for two', 'Price range', 'Aggregate rating']]

# Example: Recommend restaurants with cuisine preference 'Italian' (assuming 'Italian' corresponds to a binary column)
# Assuming 'Italian' is a column in the one-hot encoded data, for example 'French', 'Italian', 'Chinese' etc.
recommended_italian_restaurants = recommend_by_cuisine(cuisine_preference='Italian', top_n=5)
print(recommended_italian_restaurants)


      Restaurant ID Cuisines  Average Cost for two  Price range  \
706         2500023       []             -0.024764     1.319803   
9140        3700017       []             -0.018561     1.319803   
9214       18367402       []             -0.018561     1.319803   
76          6705858       []             -0.049578     2.424090   
9488        6515339       []             -0.049578     2.424090   

      Aggregate rating  
706           0.615730  
9140          0.615730  
9214          0.615730  
76            1.143331  
9488          1.143331  


#Test the Recommendation System

In [26]:
# Get 10 restaurants similar to the one with ID 5
recommended_restaurants = recommend_restaurants(restaurant_id=5, top_n=10)
print(recommended_restaurants)


      Restaurant ID Cuisines  Average Cost for two  Price range  \
2450          15774       []             -0.006154     1.319803   
2452          15321       []              0.000049     1.319803   
2285        2100870       []              0.006252     1.319803   
6709         312476       []              0.006252     1.319803   
9259        2800903       []              0.006252     1.319803   
844         3000048       []              0.012456     1.319803   
9275        2800019       []              0.012456     1.319803   
745        18339874       []              0.018659     1.319803   
3438           3192       []              0.018659     1.319803   
4386           3190       []              0.018659     1.319803   

      Aggregate rating  
2450          1.143331  
2452          1.143331  
2285          1.143331  
6709          1.143331  
9259          1.143331  
844           1.143331  
9275          1.143331  
745           1.143331  
3438          1.143331  
4386        

In [30]:
# Recommend top 5 Italian restaurants
recommended_italian_restaurants = recommend_by_cuisine(cuisine_preference='Italian', top_n=5)
print(recommended_italian_restaurants)



      Restaurant ID Cuisines  Average Cost for two  Price range  \
706         2500023       []             -0.024764     1.319803   
9140        3700017       []             -0.018561     1.319803   
9214       18367402       []             -0.018561     1.319803   
76          6705858       []             -0.049578     2.424090   
9488        6515339       []             -0.049578     2.424090   

      Aggregate rating  
706           0.615730  
9140          0.615730  
9214          0.615730  
76            1.143331  
9488          1.143331  


#Evaluate the Quality of Recommendations


Precision: The fraction of recommended restaurants that are relevant to the user.

Recall: The fraction of relevant restaurants that were recommended.

User feedback: Let users rate the recommendations (1-5 stars).
